src/redux/reducers/genes/getInitialState.js

```diff
@@ -13,12 +13,20 @@ const getInitialState = () => ({
     data: {},
   },
   expression: {
-    loading: [],
-    error: false,
-    matrix: new ExpressionMatrix(),
-    downsampledMatrix: new ExpressionMatrix(),
-    downsampledCellOrder: [],
     views: {},
+    full: {
+      loading: [],
+      error: false,
+      matrix: new ExpressionMatrix(),
+    },
+    downsampled: {
+      loading: [],
+      error: false,
+      matrix: new ExpressionMatrix(),
+      cellOrder: [],
+      downsampleSettings: null,
+      ETag: null,
```

Comment:
I added this to avoid racing conditions in many different downsample expression requests.

Each new downsampled expression work is likely to replace the previous data we are showing. This way we can coordinate to only consider the data loaded when the work result that matches this ETag (the last request) arrives.

Reply:
This isn't a problem in other work requests because users can't usually trigger a new work request while another one is already loading.

Reply:
How was this not a problem for other requests? How come we have a race condition here and not just be able to use the last loaded genes always without checking the ETag?

Reply:
Other requests usually are blocked from being called again until the currently loading one is finished.

In the case of downsampling in the heatmap, if we followed that pattern, this would mean that if you just want to move a group by from the bottom to the top (by clicking the up arrow 3 or 4 times for example), you would need to click and wait for the request to finish 3 or 4 times.

What I tried to do here was to preserve the current behavior of the buttons that modify the downsampling.


> How come we have a race condition here and not just be able to use the last loaded genes always without checking the ETag
Because the last loaded genes are for 1000 cells, and we might actually need a different 1000 cells now